### ISSUE TITLE
Failed to join cluster in Pull-Mode

### ISSUE THREAD

**Issue by maaft** [karmada-io/karmada#4262]
**What happened**:

1. karmada installed in host-cluster and api-server is publicly exposed under <domain> and is reachable and functional (can do kubectl apply  etc)
2. in on-premise cluster (private networking; but access to internet) I try to join the host-cluster:

```shell
karmadactl register <domain> --token bsbmb3.<secret> --discovery-token-ca-cert-hash sha256:bb5593ab60547232a812109d87d48d3f16834bb68fa99035e64c8beeea442dd3
[preflight] Running pre-flight checks
[prefligt] All pre-flight checks were passed
[karmada-agent-start] Waiting to perform the TLS Bootstrap
[karmada-agent-start] Waiting to construct karmada-agent kubeconfig
Unable to connect to the server: dial tcp: lookup karmada-apiserver.karmada-system.svc.cluster.local on 127.0.0.53:53: server misbehaving
```
The join fails.

**What you expected to happen**:

I expect the member cluster to be added successfully.

**Comment by RainbowMango** [comment]
> Unable to connect to the server: dial tcp: lookup karmada-apiserver.karmada-system.svc.cluster.local on 127.0.0.53:53: server misbehaving

Can you access karmada-apiserver on the machine where running the `karmadactl register`?

**Comment by maaft** [comment]
yes, I can access it (via `kubectl`) from the machine where I'm running `karmadactl register`.

I'm not sure, why karmadactl even tries to resolve `karmada-apiserver.karmada-system.svc.cluster.local` on my local DNS, since I specified that karmada-api server must be found at `<domain>`?

**Comment by maaft** [comment]
I did some further debugging:

`karmadaClusterInfo` in `pgk/karmadactl/register/register.go:340` contains this
```json
{
  "LocationOfOrigin": "",
  "server": "https://karmada-apiserver.karmada-system.svc.cluster.local:5443",
  "certificate-authority-data": "LS0tLS1C..."
}
```

Shouldn't there be `karmada-apiserver.example.com` instead?

**Comment by maaft** [comment]
Okay, I think I found the issue:

1. karmada retrieves control plane info using bootstrap token
2. this info contains `karmada-apiserver.karmada-system.svc.cluster.local` and not `karmada-apiserver.example.com` 
3. agent kubeconfig is build with that info
4. agent tries to connect with that config and fails.

[register.go -> getClusterInfoFromControlPlane](pgk/karmadactl/register/register.go#L852) should use original domain and not the domain that is retrieved during bootstrap by control-plane.

**Comment by RainbowMango** [comment]
cc @lonelyCZ @chaosi-zju for help

**Comment by liangyuanpeng** [comment]
@maaft 

This server  info is using karmada-apiserver.config, are you using `karmada-apiserver.example.com` by karmada-apiserver.config.

You can try to the blew steps:
1. change `karmada-apiserver.example.com` to `karmada-apiserver.example.com`  for karmada-apiserver.config
2. karmadactl create token with karmada-apiserver.config
3. `karmadactl registry` again

**Comment by everpeace** [comment]
`karmada-apiserver.karmada-system.svc.cluster.local` comes from `default/cluster-info` configmap in the karmada controlplane.

I think `karmada register` should not use the endpoint from cluster-info, but used the endpoint(bootstrap endpoint) which user passed to the command. 

It is because the endpoint in cluster-info is sometimes unreachable from member cluster as the issue reported. But, it can guarantee bootstrap endpoint can be reachable (generating kubeconfig step is after bootstrapping TLS (CA) certificate).

I think #4562 can be a fix.
